### PR TITLE
Generate Adversary From Layer

### PR DESCRIPTION
**Pull request by wbooth**
Create Adversary from layer file

### CODE REVIEW

**Comment by ghost** [comment]
@wbooth build failure

File: app/compass_svc.py [hunk 1]
```diff
@@ -1,15 +1,30 @@
 import json
+import uuid
 
 from aiohttp import web
 from aiohttp_jinja2 import template
 
 
+from pprint import pprint # TODO: Remove
```
Comment (ghost) [inline app/compass_svc.py]: TODO..

File: app/compass_svc.py [hunk 2]
```diff
@@ -1,15 +1,30 @@
...
+
+def check_authorization(func):
+    async def process(func, *args, **params):
```
Comment (ghost) [inline app/compass_svc.py]: I like this but if we do it-- we should make it global/in the core code.

Comment (ghost) [reply app/compass_svc.py]: like maybe it's in the auth_svc or something..

File: templates/compass.html [hunk 3]
```diff
@@ -45,6 +49,22 @@ <h4>
...
+                            <br>
+                            <div style="background-color: var(--primary-background);padding: 15px;border-radius: 25px">
+                                <h4>
+                                Create an Adversary to use in an operation. Follow the steps below to create your own
+                                    Adversary:
+                                </h4>
+                                <ol>
+                                    <li>Select Techniques: Right Click and 'add to selection' </li>
+                                    <li>Once all techniques are selected, change the score (bar chart under 'technique
+                                        controls' to a positive number</li>
```
Comment (ghost) [inline templates/compass.html:43]: This is an overwhelming amount of text when you load the page. I'd leverage the ?/duk pop ups to hide some of it and show only what you need to. or reword/restructure what you're trying to get across.

File: templates/compass.html [hunk 4]
```diff
@@ -31,6 +31,10 @@ <h2 style="margin-top:-55px">find your way</h2>
                             </div>
                             <button id="generateLayer" type="button" class="button-success"
                                     style="" onclick="generateLayer()">Generate Layer</button>
+                            <input id='adversaryLayerInput' type='file' hidden/>
+                            <button id="generateAdversary" type="button" class="button-success"
```
Comment (ghost) [inline templates/compass.html]: these buttons are too far apart. but overall, now that this modal section can do two things, you should consider redesigning the way it looks. no need to copy the 2-column look of most of the modals. this one may need something a bit more custom to look right.

File: app/compass_svc.py [hunk 5]
```diff
@@ -72,3 +76,62 @@ def _get_layer_boilerplate(name, description):
...
+        while True:
+            field = await reader.next()
+            if not field:
+                break
+            while True:
+                chunk = await field.read_chunk()
+                if not chunk:
+                    break
+                chunks.append(chunk)
+        body = b''.join(chunks)
```
Comment (ghost) [inline app/compass_svc.py]: this screams python2 to me. I think there's a more python3ic way to do this part, with the byte join

File: app/compass_svc.py [hunk 6]
```diff
@@ -72,3 +76,62 @@ def _get_layer_boilerplate(name, description):
...
+        request_body = json.loads(body)
```
Comment (ghost) [inline app/compass_svc.py]: just return here, no need for the extra variable creation

File: app/compass_svc.py [hunk 7]
```diff
@@ -72,3 +76,62 @@ def _get_layer_boilerplate(name, description):
             layer['techniques'].append(technique)
 
         return web.json_response(layer)
+
+    @staticmethod
+    def extract_techniques(request_body):
+        techniques = request_body.get('techniques')
+        adversary_techniques = set()
+        for technique in techniques:
+            if technique.get('score') > 0:
```
Comment (ghost) [inline app/compass_svc.py:85]: why do we care about the score here?

File: templates/compass.html (hunk 4 above)
Comment (wbooth) [reply templates/compass.html]: updated
Untitled

File: app/compass_svc.py (hunk 5 above)
Comment (wbooth) [reply app/compass_svc.py]: I changed to append bytes from the beginning and skip the intermediary list. But did not find a consensus on a more python3 approach. I saw PEP460: https://www.python.org/dev/peps/pep-0460/ but mentions the same approach as here

File: app/compass_svc.py (hunk 7 above)
Comment (wbooth) [reply app/compass_svc.py:85]: Score determines what technique is included. Can also include layer techniques that are colored. Score seemed the most straightforward and standard way of using layers and avoid the use of interpreting the meaning colors. (Maybe a red colored technique means DONT INCLUDE)

File: templates/compass.html (hunk 3 above)
Comment (wbooth) [reply templates/compass.html:43]: done see image below

File: app/compass_svc.py (hunk 7 above)
Comment (ghost) [reply app/compass_svc.py:85]: ah interesting.. ok. this probably justifies a comment

File: templates/compass.html [hunk 8]
```diff
@@ -6,23 +6,64 @@
...
             <div class="section-profile">
                     <div class="row">
+                        <div class="topright duk-icon"><img onclick="openHelp()" src="/gui/img/duk.png"></div>
                         <div class="column section-border" style="flex:37%;text-align:left;padding:15px;">
                             <h1 style="font-size:70px;margin-top:-20px;">Compass</h1>
-                            <h2 style="margin-top:-55px">find your way</h2>
-                            <br><br>
+                            <h2 style="margin-top:-50px">find your way</h2>
+                        </div>
+                        <div class="column section-border" style="flex:37%;text-align:left;padding:15px;">
```
Comment (ghost) [inline templates/compass.html:66]: under the sub-header (find your way) I would add a <p> that says something like "you can generate a layer file for any adversary, which you can overlay on the matrix below. Vice-versa, you can create an adversary from the matrix itself." 

something like this that will explain to a newbie what a layer is.

Comment (wbooth) [reply templates/compass.html:66]: Updated

**Pull request merged**